### PR TITLE
Refactored HpToolsLauncher, Add support for ALM SSO

### PR DESCRIPTION
**Pull request by aslaakso**
Refactored HpToolsLauncher
https://issues.jenkins-ci.org/browse/JENKINS-59006

Add support for ALM SSO (ALM 15)
https://issues.jenkins-ci.org/browse/JENKINS-59007

Bug fixes:
https://issues.jenkins-ci.org/browse/JENKINS-58175

### CODE REVIEW

File: HpToolsLauncher/Launcher.cs [hunk 1]
```diff
@@ -281,61 +272,44 @@ public void Run()
             RunTests(runner, resultsFilename, results);
 
        
-            if (_runtype.Equals(TestStorageType.Alm))
+            switch (_runType)
             {
-                bool filterSelected;
-                string filter = (_ciParams.ContainsKey("FilterTests") ? _ciParams["FilterTests"] : "");
+                case TestStorageType.Alm:
```
Comment (grount) [inline HpToolsLauncher/Launcher.cs]: Delete this and move to if statemment

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 2]
```diff
@@ -154,122 +168,351 @@ public ITDConnection2 TdConnection
...
             {
-                FindAllTestSetsUnderFolders();
+                object conn = Activator.CreateInstance(type);
+               _tdConnection = conn as ITDConnection13;
+               
             }
-            catch (Exception ex)
+            catch (FileNotFoundException ex)
             {
+                ConsoleWriter.WriteLine(GetAlmNotInstalledError());
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Assuming that WriteLine outputs the issue in the console output of the Jenkins job, it still would a good idea to log the issue.

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 3]
```diff
@@ -154,122 +168,351 @@ public ITDConnection2 TdConnection
...
+            try
+            {
+                object conn = Activator.CreateInstance(type);
+                _tdConnectionOld = conn as ITDConnection2;
+            }
+            catch (FileNotFoundException ex)
             {
-                string testset1 = testset.TrimEnd("\\".ToCharArray());
+                ConsoleWriter.WriteLine(GetAlmNotInstalledError());
+                ConsoleWriter.WriteLine(ex.Message);
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Same

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 4]
```diff
@@ -154,122 +168,351 @@ public ITDConnection2 TdConnection
...
-                            {
-                                tsName = testset1.Substring(pos, qmIndex - pos).Trim("\\".ToCharArray()).TrimEnd(' ');
-                                testParameters = testset1.Substring(qmIndex, testset1.Length - qmIndex).Trim("\\".ToCharArray());
-                            }
+                            TdConnection.Login(qcClientID, qcApiKey);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        ConsoleWriter.WriteLine(ex.Message);
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Same

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 5]
```diff
@@ -154,122 +168,351 @@ public ITDConnection2 TdConnection
...
+                        TdConnection.InitConnectionEx(qcServerUrl);
+                    }
+                    else
+                    {
+                        TdConnection.InitConnectionWithApiKey(qcServerUrl, qcClientID, qcApiKey);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    ConsoleWriter.WriteLine(ex.Message);
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Same

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 6]
```diff
@@ -154,122 +168,351 @@ public ITDConnection2 TdConnection
...
+             
+           
+            ITestSetFolder tsFolder = null;
+            try
+            {
+                tsFolder = (ITestSetFolder)tsTreeManager.get_NodeByPath(testSet);
+            }
+            catch (Exception ex)
+            {
+                return null;
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Shouldn't it be logged?

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 7]
```diff
@@ -613,232 +778,523 @@ private string GetTestInstancesString(ITestSet set)
...
+                Console.WriteLine("Null test set list");
+            }
+
+            ConsoleWriter.WriteLine(Resources.GeneralDoubleSeperator);
+            ConsoleWriter.WriteLine(Resources.AlmRunnerStartingExecution);
+            ConsoleWriter.WriteLine(string.Format(Resources.AlmRunnerDisplayTest, testSuiteName, targetTestSet.ID));
+
+            //start execution
+            ITSScheduler scheduler = null;
+            try
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Unite the try clauses

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 8]
```diff
@@ -613,232 +778,523 @@ private string GetTestInstancesString(ITestSet set)
...
+            {
+                //need to run this to install everything needed http://AlmServer:8080/qcbin/start_a.jsp?common=true
+                //start the scheduler
+                scheduler = targetTestSet.StartExecution("");
+            }
+            catch (Exception ex)
+            {
+                scheduler = null;
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: there is no need to set it to null

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 9]
```diff
@@ -613,232 +778,523 @@ private string GetTestInstancesString(ITestSet set)
...
+                        scheduler.RunAllLocally = false;
+                        break;
                 }
             }
+            catch (Exception ex)
+            {
+                ConsoleWriter.WriteLine(string.Format(Resources.AlmRunnerProblemWithHost, ex.Message));
+            }
+
+            //  ConsoleWriter.WriteLine(Resources.AlmRunnerNumTests + " " + tsList.Count);
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Remove comment

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 10]
```diff
@@ -859,24 +1315,34 @@ private string GetTestInstancesString(ITestSet set)
...
+        /// <param name="currentTest"></param>
+        /// <param name="executionStatus"></param>
+        /// <param name="targetTestSet"></param>
+        /// <param name="activeTestDesc"></param>
+        /// <param name="runDesc"></param>
+        /// <param name="testPath"></param>
+        /// <param name="abortFilename"></param>
         private void SetTestResults(ITSTest currentTest, IExecutionStatus executionStatus, ITestSet targetTestSet, TestRunResults activeTestDesc, TestSuiteRunResults runDesc, string testPath, string abortFilename)
         {
+            if (currentTest == null) throw new ArgumentNullException("currentTest");
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Please write more descriptive exception.

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 11]
```diff
@@ -859,24 +1315,34 @@ private string GetTestInstancesString(ITestSet set)
...
+
+            if (activeTestDesc == null) throw new ArgumentNullException("activeTestDesc");
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Please write more descriptive exception.

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 12]
```diff
@@ -1106,23 +1418,24 @@ private TestRunResults UpdateTestStatus(TestSuiteRunResults runResults, ITestSet
                         case TestState.Error:
                             qTest.ErrorDesc = testExecStatusObj.Status + " : " + testExecStatusObj.Message;
                             break;
+                        case TestState.Waiting:
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Is there any need for those clauses?

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 13]
```diff
@@ -1131,318 +1444,413 @@ private TestRunResults UpdateTestStatus(TestSuiteRunResults runResults, ITestSet
...
-        private string GetTestRunLog(ITSTest currentTest)
+        /// <param name="abortFilename"></param>
+        public void UpdateTesResultsDescription(ref TestRunResults activeTestDesc, TestSuiteRunResults runDesc,
+                                             ITSScheduler scheduler, ITestSet targetTestSet,
+                                             string currentTestSetInstances, double timeout,
+                                             IExecutionStatus executionStatus, Stopwatch sw,
+                                             ref ITSTest prevTest, ref ITSTest currentTest, string abortFilename)
         {
-            string TestLog = "log\\vtd_user.log";
+            // if (activeTestDesc == null) throw new ArgumentNullException(nameof(activeTestDesc));
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Remove this comment..

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 14]
```diff
@@ -1131,318 +1444,413 @@ private TestRunResults UpdateTestStatus(TestSuiteRunResults runResults, ITestSet
...
 
-            IRun lastrun = currentTest.LastRun as IRun;
-            string retVal = "";
-            if (lastrun != null)
+            // if (currentTest == null) throw new ArgumentNullException(nameof(currentTest));
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Remove this comment..

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 15]
```diff
@@ -1131,318 +1444,413 @@ private TestRunResults UpdateTestStatus(TestSuiteRunResults runResults, ITestSet
...
-            }
-            else
+            var oldQc = CheckIsOldQc();
+            var useSsl = MQcServer.Contains("https://");
+
+            string linkStr = "";
+            if (!oldQc)
             {
-                oldQc = true;
+                linkStr = useSsl ? ("tds://" + MQcProject + "." + MQcDomain + "." + MQcServer.Replace("https://", "") + "/TestRunsModule-00000000090859589?EntityType=IRun&EntityID=" + runId)
```
Comment (grount) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: What does the hardcoded string means?

File: HpToolsLauncher/TestRunners/GuiTestRunner.cs [hunk 16]
```diff
@@ -431,7 +431,11 @@ private GuiTestRunResult ExecuteQTPRun(TestRunResults testResults)
                 Type runResultsOptionstype = Type.GetTypeFromProgID("QuickTest.RunResultsOptions");
                 var options = (RunResultsOptions)Activator.CreateInstance(runResultsOptionstype);
                 options.ResultsLocation = testResults.ReportLocation;
-                _qtpApplication.Options.Run.RunMode = _uftRunMode;
+                // _qtpApplication.Options.Run.RunMode = _uftRunMode;
```
Comment (grount) [inline HpToolsLauncher/TestRunners/GuiTestRunner.cs]: Remove this comment..

File: src/main/java/com/microfocus/application/automation/tools/model/SseModel.java [hunk 17]
```diff
@@ -201,4 +208,49 @@ public String getPostRunAction() {
...
+        props.put("almUserName", getAlmUserName());
+
+        String encAlmPass = "";
+        try {
+
+            encAlmPass =
+                    EncryptionUtils.Encrypt(
+                            getAlmPassword(),
+                            EncryptionUtils.getSecretKey());
+        }catch (Exception e){
```
[the remark this answers is not in the document]
Comment (grount) [reply src/main/java/com/microfocus/application/automation/tools/model/SseModel.java]: Use the builtin jenkins logger

File: src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java [hunk 18]
```diff
@@ -180,13 +184,85 @@ public void perform(Run<?, ?> build, FilePath workspace, Launcher launcher,
...
+        String propsSerialization = stream.toString();
+        InputStream propsStream = IOUtils.toInputStream(propsSerialization);
+
+        // get the remote workspace filesys
+        FilePath projectWS = workspace;
+
+        // Get the URL to the Script used to run the test, which is bundled
+        // in the plugin
+        URL cmdExeUrl =
+                Hudson.getInstance().pluginManager.uberClassLoader.getResource(HP_TOOLS_LAUNCHER_EXE);
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java]: Hudson is a legacy library, why not use Jenkins?

File: src/main/resources/com/microfocus/application/automation/tools/run/AutEnvironmentBuilder/config.jelly [hunk 19]
```diff
@@ -23,6 +23,40 @@
...
 
+    <style>
+         #testName {
+               width: 300px;
+               margin-bottom: 10px;
+           }
+
+           table tr td div.help {
+               margin-left: 12px;
+               background: none !important;
```
Comment (grount) [inline src/main/resources/com/microfocus/application/automation/tools/run/AutEnvironmentBuilder/config.jelly]: why do you use !important everywhere?

File: src/main/java/com/microfocus/application/automation/tools/sse/sdk/RunManager.java [hunk 20]
```diff
@@ -144,6 +144,7 @@ private void setRunId(RunResponse runResponse) {
      * Log report url
      */
     private void logReportUrl(boolean isSucceeded, Args args) {
+        isSucceeded = true;
```
[the remark this answers is not in the document]
Comment (aslaakso) [reply src/main/java/com/microfocus/application/automation/tools/sse/sdk/RunManager.java]: Removed the assignment. It was just for testing one use case.

File: src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java (hunk 18 above)
Comment (aslaakso) [reply src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java]: I guess that was the initial implementation when the construction of the plugin started.

Comment (aslaakso) [reply src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java]: Changed to Jenkins

File: HpToolsLauncher/TestRunners/GuiTestRunner.cs (hunk 16 above)
Comment (aslaakso) [reply HpToolsLauncher/TestRunners/GuiTestRunner.cs]: Removed

File: src/main/resources/com/microfocus/application/automation/tools/run/AutEnvironmentBuilder/config.jelly (hunk 19 above)
Comment (aslaakso) [reply src/main/resources/com/microfocus/application/automation/tools/run/AutEnvironmentBuilder/config.jelly]: Removed

File: src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java [hunk 21]
```diff
@@ -180,13 +185,100 @@ public void perform(Run<?, ?> build, FilePath workspace, Launcher launcher,
...
+            PrintStream out = listener.getLogger();
+            try {
+                AlmToolsUtils.runHpToolsAborterOnBuildEnv(build, launcher, listener, ParamFileName, workspace);
+            } catch (IOException e1) {
+                Util.displayIOException(e1, listener);
+                build.setResult(Result.FAILURE);
+                return;
+            } catch (InterruptedException e1) {
+                // TODO Auto-generated catch block
+                e1.printStackTrace();
```
[the remark this answers is not in the document]
Comment (aslaakso) [reply src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java]: logger added

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs (hunk 15 above)
Comment (aslaakso) [reply HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: That is the link that opens the results in ALM

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs (hunk 8 above)
Comment (aslaakso) [reply HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: removed

Comment (aslaakso) [reply HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: removed

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs (hunk 11 above)
Comment (aslaakso) [reply HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: updated

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs (hunk 13 above)
Comment (aslaakso) [reply HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: removed

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs (hunk 7 above)
Comment (aslaakso) [reply HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: united

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 22]
```diff
@@ -154,122 +169,351 @@ public ITDConnection2 TdConnection
...
+            try
+            {
+                object conn = Activator.CreateInstance(type);
+                _tdConnectionOld = conn as ITDConnection2;
+            }
+            catch (FileNotFoundException ex)
             {
-                string testset1 = testset.TrimEnd("\\".ToCharArray());
+                ConsoleWriter.WriteLine(GetAlmNotInstalledError());
+                ConsoleWriter.WriteLine(ex.Message);
```
Comment (NarcisaMGalan) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Did you make the change here ?

File: HpToolsLauncher/Runners/AlmTestSetsRunner.cs [hunk 23]
```diff
@@ -154,122 +169,351 @@ public ITDConnection2 TdConnection
...
+
+                    if (TdConnection.LoggedIn)
+                    {
+                        try
+                        {
+                            TdConnection.Connect(qcDomain, qcProject);
+                        }
+                        catch (Exception ex)
+                        {
+                            Console.WriteLine(ex.Message);
```
Comment (NarcisaMGalan) [inline HpToolsLauncher/Runners/AlmTestSetsRunner.cs]: Shouldn't you use logger?

File: src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java [hunk 24]
```diff
@@ -180,13 +184,84 @@ public void perform(Run<?, ?> build, FilePath workspace, Launcher launcher,
...
+        try {
+            // create a file for the properties file, and save the properties
+            propsFileName.copyFrom(propsStream);
+
+            // Copy the script to the project workspace
+            CmdLineExe.copyFrom(cmdExeUrl);
+        } catch (IOException e1) {
+            build.setResult(Result.FAILURE);
+            // TODO Auto-generated catch block
+            e1.printStackTrace();
```
Comment (NarcisaMGalan) [inline src/main/java/com/microfocus/application/automation/tools/run/SseBuilder.java]: Logger?

File: src/main/webapp/css/alm.css [hunk 25]
```diff
@@ -23,10 +23,29 @@
...
+    background: none !important;
+    background-color: transparent !important;
+}
+
+table td .help {
+    background: none !important;
+}
+
+table tr[nameref^="rowSetStart"] {
+    background: none !important;
```
Comment (NarcisaMGalan) [inline src/main/webapp/css/alm.css]: Do you need "!important" here ?

File: src/main/webapp/css/build_tasks.css [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+    background: none !important;
+    background-color: transparent;
+}
+
+table td.help {
+    background: none !important;
+}
+
+table tr[nameref^="rowSetStart"] {
+    background: none !important;
```
Comment (NarcisaMGalan) [inline src/main/webapp/css/build_tasks.css]: Also here?

File: src/main/webapp/css/alm.css (hunk 25 above)
Comment (aslaakso) [reply src/main/webapp/css/alm.css]: removed

File: src/main/webapp/css/build_tasks.css (hunk 26 above)
Comment (aslaakso) [reply src/main/webapp/css/build_tasks.css]: removed

**Pull request merged**